File: python/script_sim_suh.py
# %%
import numpy as np

from src.ctrldesign.ode_modelling.discrete_model import discretize_model
from src.ctrldesign.ode_modelling.ode_model import get_linearized_matrices, get_model
from src.ctrldesign.ode_modelling.simresult import plot_sim_results
from src.ctrldesign.ode_modelling.simulator import sim_nl
from src.model_suh import model_suh

#! %matplotlib qt

# %% Load model with standard parameters

p = model_suh.params()
model = get_model(model_suh.model, p)


# %% Simulate continuous model
Ts = 0.01
(u, x0) = model_suh.get_validation_input(Ts)
u = u[:, : int(25 // Ts)]

res = sim_nl(model, Ts, x0, u)
res.desc = f"cont., Ts = {Ts} s"

plot_sim_results(res, signal_infos=model_suh.signal_infos())


# %% Simulate discretized model

Ts = 0.001
dmodel = discretize_model(model, Ts, "rk4")

(u, x0) = model_suh.get_validation_input(Ts)
u = u[:, : int(25 // Ts)]

dres = sim_nl(dmodel, Ts, x0, u)
dres.desc = f"disc., Ts = {Ts} s"

plot_sim_results((dres, res), signal_infos=model_suh.signal_infos())


# %% Have a look at the eigenvalues of the linearized system

u_pre_change_idx = set(np.where(res.u[0, 1:] != res.u[0, :-1])[0])
u_pre_change_idx.union(set(np.where(res.u[1, 1:] != res.u[1, :-1])[0].flatten()))

u_pre_change_idx = sorted(u_pre_change_idx)

scaled = True

for idx in u_pre_change_idx:

    tk = res.t[idx]
    xk = res.x[:, idx]
    uk = res.u[:, idx]

    (A, _, _, _) = get_linearized_matrices(model, xk, uk)

    evs = np.linalg.eigvals(A)
    evs.sort()

    print(f"eigenvalues of A evaluated at (x({tk:.2f}), u({tk:.2f})):", evs)

File: matlab/script_sim_suh.m
%%
clc();
init_paths();

%%
p = params_suh();
model = get_model(@model_suh, p);

%%

Ts = 0.01;
[u, x0] = get_validation_input_suh(Ts);

res = sim_nl(model, Ts, x0, u);
res.desc = sprintf('continuous, Ts=%g s', Ts);
plot_sim_result(res, 'SignalInfos', get_signal_infos_suh());

%%
[A, B, C, D] = get_linearized_matrices(model, x0, u(:, 1));

disp('Eigenwerte von A:')
disp(eig(A));

disp('1 / |lambda(A)|:')
disp(1 ./ abs(eig(A)));

%%

Ts = 0.001;
[u, x0] = get_validation_input_suh(Ts);

dmodel = discretize_model(model, Ts, 'rk4');
dres = sim_nl(dmodel, Ts, x0, u);
dres.desc = sprintf('discrete, Ts=%g s', Ts);
plot_sim_result(dres, 'SignalInfos', get_signal_infos_suh());

%%

plot_sim_result({res, dres}, 'ReuseFigure', true, 'SignalInfos', get_signal_infos_suh());


%%

n_step = round(length(res.t) / 100);

idx = (1:n_step:length(res.t) - 1);
evs = nan(length(model.states), length(idx));
t_evs = res.t(idx);

for i = 1:length(idx)
    [A, B, C, D] = get_linearized_matrices(model, res.x(:, idx(i)), u(:, idx(i)));

    evs(:, i) = eig(A);
end

%%

vh = subplot(2, 1, 1);
plot(t_evs, real(evs).');
ylabel('real(ev)');

vh(2) = subplot(2, 1, 2);
plot(t_evs, 1 ./ abs(real(evs).'));
ylabel('1 / abs(ev)');
xlabel('time');
linkaxes(vh, 'x');
